src/core/app.lua

```diff
@@ -124,6 +128,65 @@ function configure (new_config)
    counter.add(configs)
 end
 
+-- Claims a name for a program so it can be identified by name by other processes.
+--
+-- The name given to the function must be unique, if a name has been used before by
+-- an active process the function will error displaying an appropriate error message.
+-- Any other process can enumerate programs by their name and PID using enumerate_named_programs.
+function claim_name(name)
+   configuration[name] = name
+   local namedir = "by-name/" .. name
+   local namedirqualified = named_program_root .. "/" .. name
+   local piddir = shm.root .. "/" .. S.getpid()
+
+   -- Verify that the by-name directory exists.
+   shm.mkdir(namedir)
+   
+   -- Check if a symlink to a named program with the same name exists, if not, unlink it.
+   if S.lstat(namedirqualified) then
+      -- It exists, lets check what the PID is.
+      local piddir = S.readlink(namedirqualified)
+      local ps, pe = string.find(piddir, "/[^/]*$")
+      local pid = tonumber(string.sub(piddir, ps+1, pe))
+      if S.kill(pid, 0) then
+         -- It's a running process, we should display an error message.
+         error("Name has been claimed by a currently running process: "..namedirqualified)
+      else
+         -- It's dead, we should be able to remove the symlink and relink.
+         S.unlink(namedirqualified)
+      end
+   end
+   
+   -- Create the new symlink.
+   local rtn, err = S.symlink(piddir, namedirqualified)
+   if rtn == -1 then
+      error("Error creating program name symlink in: "..namedirqualified)
+   end
+end
+
+-- Enumerates all the named programs with their PID
+--
+-- This returns a table programs with the key being the name of the program
+-- and the value being the PID of the program. Each program is checked that
+-- it's still alive. Any dead program or program without a name is not listed.
+function enumerate_named_programs()
+   local progs = {}
+   local dirs = S.util.dirtable(named_program_root, true)
```

Comment:
Is this the only instance of S.util in this file?  If so probably we shouldn't use it.  Does `core.lib.files_in_directory` work for you?  (Though, ugh, shelling out to ls!!!!!)